In [1]:
import numpy as np   #Importing Numpy Module and alaising as np 

#### Automatic generation of Numbers in array

In [2]:
a = np.arange(10) #Default 0 is the starting value and stop value is given as 10 in the example
a

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [4]:
a= np.arange(10,20,2) #Start=10 End =20 (Excluded) , Jump=2
a

array([10, 12, 14, 16, 18])

#### Why Numpy is better than lists?

In [5]:
L = range(1000) #Created a List with 1000 elements from 0 to 999
%timeit [i**2 for i in L] #timeit is magical function that is available in Jupyter notebook

832 µs ± 34.4 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


So we got 832µs per loop while doing multiplication by 2 for each element in list of 100 elements

In [6]:
L = np.arange(1000) #Created a Numpy array with 1000 elements from 0 to 999
%timeit L**2

3.61 µs ± 108 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


It just took 3.61µs per loop which is around 250 times better

#### Creating A Array

In [7]:
a= np.array([3,5,6,1,7]) #Creating a Numpy array with List elements
a

array([3, 5, 6, 1, 7])

In [8]:
a.ndim #Will return the Number of Dimension (It is 1 dimension and it is scalar)

1

In [9]:
a.shape #Since it is only 1 dimension and it returns the number of elements in that array, will understand its use better in 2D array

(5,)

In [10]:
len(a) #Will return the number of elements in the list

5

#### Create a 2D array

In [16]:
a = np.array([[2,3,5],[3,4,5]]) # Creating a 2D Array/Matrix # It is like List of list
a

array([[2, 3, 5],
       [3, 4, 5]])

In [17]:
type(a) #Checking the datatype of this array - n dimensional array

numpy.ndarray

In [18]:
a.ndim #It is 2 dimension array

2

In [19]:
a.shape #For 2 dimension arrays shape will return rows,columns , Here rows=2, columns=3

(2, 3)

In [20]:
len(a) #It will return the number of rows

2

#### Create a 3D array

In [41]:
a = np.array([            #Creating 3 Dimension array
    [[0,1,2],[1,2,3]],
    [[3,4,5],[5,6,7]]
    ])
a

array([[[0, 1, 2],
        [1, 2, 3]],

       [[3, 4, 5],
        [5, 6, 7]]])

In [42]:
a.ndim 

3

In [44]:
a.shape #It is of 2 breadth (2), then rows,columns

(2, 2, 3)

#### Functions for creating arrays

In [45]:
a= np.arange(2,15,3) #start, end(excluded),step-Index
a

array([ 2,  5,  8, 11, 14])

In [47]:
# Using linspace

b=np.linspace(1,3,6) #Return 6 points in between 1 and 3

b

array([1. , 1.4, 1.8, 2.2, 2.6, 3. ])

In [49]:
c= np.ones((2,2)) # Creates a 2 * 2 matrix  with all values as 1
c

array([[1., 1.],
       [1., 1.]])

In [50]:
d= np.zeros((2,2)) # Creates a 2 * 2 matrix  with all values as 0
d

array([[0., 0.],
       [0., 0.]])

In [52]:
e=np.eye(3) #All the diagonal numbers are filled with value 1. It is unit matrix or Identity Matrix
e

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [54]:
f=np.eye(3,2) # 1*1, 2*2, 3*3 will have value 1 for Identity Matrix
f

array([[1., 0.],
       [0., 1.],
       [0., 0.]])

In [56]:
#create Array Using diagonal function

a= np.diag([1,2,4,5])

a

array([[1, 0, 0, 0],
       [0, 2, 0, 0],
       [0, 0, 4, 0],
       [0, 0, 0, 5]])

In [57]:
#extracting Diagonal Elements

np.diag(a)

array([1, 2, 4, 5])

In [58]:
#Create Array with random numbers

a=np.random.rand(4) #Creates a random array with elements between 0 and 1

a #These are uniform

array([0.94937761, 0.92391463, 0.93707447, 0.62910368])

In [62]:
a= np.random.randn(4)

a # These are probability and statistics

array([-0.84917265, -1.54297681, -0.16968304,  1.93234239])

#### Datatypes

In [64]:
a = np.arange(10)

print(a)

print(a.dtype)

[0 1 2 3 4 5 6 7 8 9]
int32


In [66]:
#We can change the dtype explicitly

a=np.arange(10,dtype=float)

print(a)

print(a.dtype)

[0. 1. 2. 3. 4. 5. 6. 7. 8. 9.]
float64


In [67]:
np.ones((3,3)) # will always create with float numbers

array([[1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.]])

In [68]:
np.ones((3,3),dtype='int64')

array([[1, 1, 1],
       [1, 1, 1],
       [1, 1, 1]], dtype=int64)

In [70]:
#Real and Imaginary Numbers

a = np.array([1+2j,3+5j])

a.dtype

dtype('complex128')

In [72]:
#Create Boolean Numpy array

b= np.array([True,False,False])

b.dtype

dtype('bool')

In [73]:
#Create String Numpy Array

c = np.array(['Santhosh','Vijay'])

c.dtype

dtype('<U8')

#### Indexing and Slicing

##### Indexing

In [75]:
a = np.arange(10)
print(a)
a[3] # Will return the value with Index 3 . Index starts from 0

[0 1 2 3 4 5 6 7 8 9]


3

In [81]:
a= np.diag([1,2,3,5])

print(a)

a[1][1] # will return the 2nd row and 2 column intersection value

[[1 0 0 0]
 [0 2 0 0]
 [0 0 3 0]
 [0 0 0 5]]


2

In [82]:
#We can assing the value also

a[1][1]=20

a #Now the value 2 is replaced with 20

array([[ 1,  0,  0,  0],
       [ 0, 20,  0,  0],
       [ 0,  0,  3,  0],
       [ 0,  0,  0,  5]])

##### Slicing

In [83]:
a = np.arange(10)
a[2:5] #Will return elements from Index 2 to 4

array([2, 3, 4])

In [84]:
a[2:8:2] #Returns the element from Index 2 to 7 with Step Index of 2

array([2, 4, 6])

In [85]:
a[::-1] #Returns all the element with step index of -1 i.e Reverse

array([9, 8, 7, 6, 5, 4, 3, 2, 1, 0])

In [86]:
#We can also assign values to multiple elements

a[3:6] = 50 #Will assign the value 50 to Elements from Index 3 to 5

a

array([ 0,  1,  2, 50, 50, 50,  6,  7,  8,  9])

In [87]:
b = np.arange(5)

a[5:]=b[::-1] # So Reversing the b array and assign to a array from index 5 to last

a

array([ 0,  1,  2, 50, 50,  4,  3,  2,  1,  0])

So you can see the last 5 values are  4,  3,  2,  1,  0 which is reverse of array b

#### Copies and Views

In [89]:
a=np.arange(10)

a

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [90]:
b= a[0:5] ## B array will hold the values of index from 0 to 4

b

array([0, 1, 2, 3, 4])

In [91]:
#Now modifying b array

b[0]=100

b

array([100,   1,   2,   3,   4])

In [92]:
#Now lets check a array

a #Here a[0] is changed to value 100

array([100,   1,   2,   3,   4,   5,   6,   7,   8,   9])

Numpy produces a View and not a copy when you are storing a slicing value to another array.
Numpy does that for memory efficiency.

In [93]:
np.shares_memory(a,b) #Will check whether both array saves in a same place

True

In [94]:
#Since a and b are pointing to the same location, I will modify values of a to checker whether it is reflecting in b

a[1]=50

b

array([100,  50,   2,   3,   4])

Yes it gets modified

In [95]:
#how to avoid this
#We can use copy explicity to make a copy and not a view

c = a[0:5].copy()

c

array([100,  50,   2,   3,   4])

In [96]:
a[0]=250

c #It doesn't reflected

array([100,  50,   2,   3,   4])

In [97]:
np.shares_memory(a,c)

False

#### Fancy Indexing

In [115]:
a= np.random.randint(0,20,15) #Will return 15 numbers between values 0 and 15
a

array([10,  3,  5,  3, 11,  6, 10,  8, 19,  3,  5, 15,  8,  6, 12])

In [116]:
#For Example, I want to retrive all the even elements in the array

a%2==0  #Will return whether each elements is odd or Even . Here True represents the number is even


array([ True, False, False, False, False,  True,  True,  True, False,
       False, False, False,  True,  True,  True])

In [117]:
a[a%2==0] #This will return the even numbers in the array

array([10,  6, 10,  8,  8,  6, 12])

In [118]:
a[a%2==0]=1 #It will modify all the even values to number 1

In [119]:
a

array([ 1,  3,  5,  3, 11,  1,  1,  1, 19,  3,  5, 15,  1,  1,  1])

In [120]:
a=np.arange(0,100,10)
a

array([ 0, 10, 20, 30, 40, 50, 60, 70, 80, 90])

In [122]:
a[[2,3,6,7]] #Will return 2,3,6 and 7 th index element

#When giving multiple indexing please make sure in giving in list of list

array([20, 30, 60, 70])

In [124]:
a[[7,8]]=400 #Modifying 7th and 8th index element to 400
a

array([  0,  10,  20,  30,  40,  50,  60, 400, 400,  90])

# ElementWise Operation

In [2]:
a=np.arange(4)

In [3]:
print(a)
print(a+1) #  Elementwise it has added by 1

[0 1 2 3]
[1 2 3 4]


In [5]:
a=a**2
print(a**2)

[ 0  1 16 81]


In [7]:
b= np.ones((3,3))
print(b)

[[1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]]


In [8]:
print(b+1)

[[2. 2. 2.]
 [2. 2. 2.]
 [2. 2. 2.]]


In [9]:
b=np.ones(4)+3
print(b)

[4. 4. 4. 4.]


In [10]:
print(a-b)
print(a+b)

[-4. -3.  0.  5.]
[ 4.  5.  8. 13.]


# Matrix Muplication

In [13]:
c=np.diag([1,2,3,4])
print(c)

[[1 0 0 0]
 [0 2 0 0]
 [0 0 3 0]
 [0 0 0 4]]


In [14]:
print(c*c)

[[ 1  0  0  0]
 [ 0  4  0  0]
 [ 0  0  9  0]
 [ 0  0  0 16]]


In [15]:
print(c.dot(c)) #same operation as c*c

[[ 1  0  0  0]
 [ 0  4  0  0]
 [ 0  0  9  0]
 [ 0  0  0 16]]


In [18]:
a=np.array([1,2,3,4])
b=np.array([3,5,3,2])

In [19]:
a==b

array([False, False,  True, False])

In [20]:
a<b

array([ True,  True, False, False])

In [22]:
#Array wise comparison

c=np.array([1,2,3,4])

In [23]:
np.array_equal(a,b)

False

In [25]:
np.array_equal(a,c)

True

# Logical Operation

In [26]:
a=np.array([0,0,1,1])
b=np.array([0,1,0,1])

In [27]:
np.logical_and(a,b)

array([False, False, False,  True])

In [30]:
np.logical_or(a,b)

array([False,  True,  True,  True])

# Transcedental Function

In [31]:
a=np.arange(5)

In [32]:
np.sin(a)

array([ 0.        ,  0.84147098,  0.90929743,  0.14112001, -0.7568025 ])

In [33]:
np.log(a) #Log 0 is infinity

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: RuntimeWarning: divide by zero encountered in log
  """Entry point for launching an IPython kernel.


array([      -inf, 0.        , 0.69314718, 1.09861229, 1.38629436])

In [34]:
np.exp(a)

array([ 1.        ,  2.71828183,  7.3890561 , 20.08553692, 54.59815003])

# Shape Mismatch

In [38]:
a= np.arange(4)
print(a)

[0 1 2 3]


In [39]:
a+np.array([1,2])

ValueError: operands could not be broadcast together with shapes (4,) (2,) 

# Basic Reductions

In [40]:
a=np.array([1,2,4,5])

In [41]:
a.sum()

12

In [42]:
a.min()

1

In [43]:
a.max()

5

In [46]:
a.argmin() #Index of Minimum value of element

0

In [47]:
a.argmax() #Index of Maximum value of element

3

In [48]:
b=np.array([[1,2],[3,4]])
print(b)

[[1 2]
 [3 4]]


In [49]:
b.sum(axis=0) #column wise addition

array([4, 6])

In [50]:
b.sum(axis=1) #Row wise addition

array([3, 7])

In [52]:
np.all([True,False])

False

In [54]:
np.any([True,False])

True

In [57]:
a=np.zeros(10)
print(a)

[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]


In [58]:
np.any(a!=0)

False

In [60]:
np.all(a==0)

True

In [62]:
a=np.array([1,2,3,2])
b=np.array([2,2,3,2])
c=np.array([6,4,4,5])
((a<=b)&(b<=c)).all() #we can try combinations of logical conditions.

True

# Statistics 

In [63]:
x=np.array([1,2,3,1])
y=np.array([[1,2,3],[4,5,6]])

In [72]:
print(x)
print(f'Mean: {x.mean()}')
print(f'Median: {np.median(x)}')

[1 2 3 1]
Mean: 1.75
Median: 1.5


In [73]:
print(y)
y.mean(axis=0)

[[1 2 3]
 [4 5 6]]


array([2.5, 3.5, 4.5])

In [75]:
np.median(y,axis=1)

array([2., 5.])

In [76]:
x.std()

0.82915619758885

# Example

In [77]:
data=np.loadtxt(r'C:/Users/Santhosh/GitHub/MachineLearning/Numpy/populations.txt')

In [78]:
data

array([[ 1900., 30000.,  4000., 48300.],
       [ 1901., 47200.,  6100., 48200.],
       [ 1902., 70200.,  9800., 41500.],
       [ 1903., 77400., 35200., 38200.],
       [ 1904., 36300., 59400., 40600.],
       [ 1905., 20600., 41700., 39800.],
       [ 1906., 18100., 19000., 38600.],
       [ 1907., 21400., 13000., 42300.],
       [ 1908., 22000.,  8300., 44500.],
       [ 1909., 25400.,  9100., 42100.],
       [ 1910., 27100.,  7400., 46000.],
       [ 1911., 40300.,  8000., 46800.],
       [ 1912., 57000., 12300., 43800.],
       [ 1913., 76600., 19500., 40900.],
       [ 1914., 52300., 45700., 39400.],
       [ 1915., 19500., 51100., 39000.],
       [ 1916., 11200., 29700., 36700.],
       [ 1917.,  7600., 15800., 41800.],
       [ 1918., 14600.,  9700., 43300.],
       [ 1919., 16200., 10100., 41300.],
       [ 1920., 24700.,  8600., 47300.]])

In [80]:
data.T #Transpose will convert rows to columns

array([[ 1900.,  1901.,  1902.,  1903.,  1904.,  1905.,  1906.,  1907.,
         1908.,  1909.,  1910.,  1911.,  1912.,  1913.,  1914.,  1915.,
         1916.,  1917.,  1918.,  1919.,  1920.],
       [30000., 47200., 70200., 77400., 36300., 20600., 18100., 21400.,
        22000., 25400., 27100., 40300., 57000., 76600., 52300., 19500.,
        11200.,  7600., 14600., 16200., 24700.],
       [ 4000.,  6100.,  9800., 35200., 59400., 41700., 19000., 13000.,
         8300.,  9100.,  7400.,  8000., 12300., 19500., 45700., 51100.,
        29700., 15800.,  9700., 10100.,  8600.],
       [48300., 48200., 41500., 38200., 40600., 39800., 38600., 42300.,
        44500., 42100., 46000., 46800., 43800., 40900., 39400., 39000.,
        36700., 41800., 43300., 41300., 47300.]])

In [81]:
years,hare,lynx,carrot= data.T

In [85]:
#Excude Year from data
populations = data[:,1:]
populations

array([[30000.,  4000., 48300.],
       [47200.,  6100., 48200.],
       [70200.,  9800., 41500.],
       [77400., 35200., 38200.],
       [36300., 59400., 40600.],
       [20600., 41700., 39800.],
       [18100., 19000., 38600.],
       [21400., 13000., 42300.],
       [22000.,  8300., 44500.],
       [25400.,  9100., 42100.],
       [27100.,  7400., 46000.],
       [40300.,  8000., 46800.],
       [57000., 12300., 43800.],
       [76600., 19500., 40900.],
       [52300., 45700., 39400.],
       [19500., 51100., 39000.],
       [11200., 29700., 36700.],
       [ 7600., 15800., 41800.],
       [14600.,  9700., 43300.],
       [16200., 10100., 41300.],
       [24700.,  8600., 47300.]])

In [87]:
#which species has highest population each year?

#here we can try max by row wise

np.argmax(populations,axis=1)

array([2, 2, 0, 0, 1, 1, 2, 2, 2, 2, 2, 2, 0, 0, 0, 1, 2, 2, 2, 2, 2],
      dtype=int64)

# BroadCasting

In [92]:
#adding a 4*3 matrix with 4*3 matrix

a= np.arange(12).reshape(4,3)
print(a)

[[ 0  1  2]
 [ 3  4  5]
 [ 6  7  8]
 [ 9 10 11]]


In [90]:
b=np.arange(12,24).reshape(4,3)
print(b)

[[12 13 14]
 [15 16 17]
 [18 19 20]
 [21 22 23]]


In [91]:
print(a+b)

[[12 14 16]
 [18 20 22]
 [24 26 28]
 [30 32 34]]


In [93]:
#adding a 4*3 matrix with 1*3 matrix

c = np.arange(20,23)
print(c)

[20 21 22]


In [94]:
print(a+c)  #1*3 matrix will be repeated 4 times to get 4*3

[[20 22 24]
 [23 25 27]
 [26 28 30]
 [29 31 33]]


In [95]:
#Adding a 4*1 matrix with 1*3

a=np.arange(4).reshape(4,1)
print(a)

[[0]
 [1]
 [2]
 [3]]


In [97]:
b=np.arange(3)
print(b)

[0 1 2]


In [99]:
print(a+b) #4*1 will try to repeat 3 times to get 4*3 shape, whereas 1*3 will repeat 4 times to get 4*3 shape.

[[0 1 2]
 [1 2 3]
 [2 3 4]
 [3 4 5]]


In [101]:
a=np.tile(np.arange(10),(3,1))
print(a)

[[0 1 2 3 4 5 6 7 8 9]
 [0 1 2 3 4 5 6 7 8 9]
 [0 1 2 3 4 5 6 7 8 9]]


In [102]:
a=np.tile(np.arange(10),(3,2))
print(a)

[[0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9]
 [0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9]
 [0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9]]


# Creating New Axis

In [115]:
a=np.arange(0,40,10)
print(a)
print(a.ndim)
print(a.shape) #1_D array

[ 0 10 20 30]
1
(4,)


In [116]:
a=a[:,np.newaxis]
print(a)
print(a.ndim)
print(a.shape) 

[[ 0]
 [10]
 [20]
 [30]]
2
(4, 1)


In [121]:
b=np.array([1,2,3])

In [111]:
a

array([ 0, 10, 20, 30])

In [122]:
a+b

array([[ 1,  2,  3],
       [11, 12, 13],
       [21, 22, 23],
       [31, 32, 33]])

# Array Shape Manipulation

In [123]:
a=np.array([[1,2,3],[4,5,6]])
print(a)

[[1 2 3]
 [4 5 6]]


In [124]:
a.ravel() #It will flatten the array. Convert Any dimension array to 1_D array

array([1, 2, 3, 4, 5, 6])

In [126]:
a.ravel().shape

(6,)

In [127]:
a.T

array([[1, 4],
       [2, 5],
       [3, 6]])

In [129]:
a=a.T.ravel()
print(a)

[1 4 2 5 3 6]


In [132]:
#The opposite of flatenning is reshaping
a.reshape(3,2).T

array([[1, 2, 3],
       [4, 5, 6]])

In [134]:
b=a.reshape(3,2)

In [136]:
np.shares_memory(a,b) #It will be created as view mostly (not all time). Check carefully

True

In [142]:
a=np.zeros((3,2))
print(a)

[[0. 0.]
 [0. 0.]
 [0. 0.]]


In [143]:
b=a.T.reshape(3,2)
print(b)

[[0. 0.]
 [0. 0.]
 [0. 0.]]


In [144]:
np.shares_memory(a,b) #In this case. it is created as copy.

False

# Dimesion Shuffling

In [145]:
a= np.arange(24).reshape(4,2,3) #Repeat 2*3 matrix 4 times
print(a)

[[[ 0  1  2]
  [ 3  4  5]]

 [[ 6  7  8]
  [ 9 10 11]]

 [[12 13 14]
  [15 16 17]]

 [[18 19 20]
  [21 22 23]]]


In [148]:
print(a[1,1,2])

11


In [160]:
a=np.arange(4)

In [161]:
a.resize((8),)

In [163]:
print(a) #It will add trailing zeros post 4, you dont need to assign this to a, resize will modify the existing array

[0 1 2 3 0 0 0 0]


In [164]:
a.resize((10,))
print(a)

[0 1 2 3 0 0 0 0 0 0]


In [167]:
a.resize((2,)) #Can extend or shrink as well
print(a)

[0 1]


In [168]:
b=a

In [169]:
print(b)

[0 1]


In [170]:
np.shares_memory(a,b)

True

In [171]:
b.resize((6,))

ValueError: cannot resize an array that references or is referenced
by another array in this way.  Use the resize function

In [172]:
a.resize((4,)) #We cannot moditfy if the array as view to another array. It should be independent to resize

ValueError: cannot resize an array that references or is referenced
by another array in this way.  Use the resize function

# Sorting

In [181]:
a=np.array([[3,1,2],[8,2,6]])
print(a)

[[3 1 2]
 [8 2 6]]


In [178]:
a.sort(axis=1)
print(a)

[[1 2 3]
 [2 6 8]]


In [179]:
a.sort(axis=0)
print(a)

[[1 2 3]
 [2 6 8]]


In [183]:
a=np.array([8,3,1,17])

In [185]:
j=np.argsort(a) #index sort of a
print(j) 

[2 1 0 3]


In [186]:
a[j]

array([ 1,  3,  8, 17])